src/app/credExplorer/state.js

```diff
@@ -0,0 +1,247 @@
+// @flow
+
+import deepEqual from "lodash.isequal";
+
+import * as NullUtil from "../../util/null";
+import {Graph} from "../../core/graph";
+import type {Repo} from "../../core/repo";
+import {type EdgeEvaluator} from "../../core/attribution/pagerank";
+import {
+  type PagerankNodeDecomposition,
+  type PagerankOptions,
+  pagerank,
+} from "../../core/attribution/pagerank";
+
+import type {DynamicPluginAdapter} from "../pluginAdapter";
+import {StaticPluginAdapter as GitAdapter} from "../../plugins/git/pluginAdapter";
+import {StaticPluginAdapter as GithubAdapter} from "../../plugins/github/pluginAdapter";
+
+/*
+  This models the UI states of the credExplorer/App as a state machine.
+
+  The different states are all instances of AppState, and the transitions are
+  explicitly managed by the StateTransitionMachine class. All of the
+  transitions, including error cases, are thoroughly tested.
+ */
+
+export type AppState = Uninitialized | Initialized;
+export type Uninitialized = {|
+  +type: "UNINITIALIZED",
+  edgeEvaluator: ?EdgeEvaluator,
+  repo: ?Repo,
+|};
+export type Initialized = {|
+  +type: "INITIALIZED",
+  +edgeEvaluator: EdgeEvaluator,
+  +repo: Repo,
+  +substate: AppSubstate,
+|};
+
+export type LoadingState = "NOT_LOADING" | "LOADING" | "FAILED";
```

Comment:
It seems like we will want to display some kind of failure message to
the user, so eventually we may have:

```js
export type LoadingState =
  | {|+type: "NOT_LOADING"|}
  | {|+type: "LOADING"|}
  | {|+type: "FAILED", +message: string|};
```

Presuming that you agree, would you like to make this change now, or
leave it for a subsequent commit as a test of how extensible this system
is? :-)

Reply:
I think it's a reasonable idea, and we can implement it later once we actually want to populate the field. (For now, you can just check the console when something goes wrong.)